object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionAnnotatedTypeFactory.java

```diff
@@ -279,6 +306,276 @@ private static String getExactStringValue(Tree tree, ValueAnnotatedTypeFactory f
     return null;
   }
 
+  @Override
+  public void postAnalyze(ControlFlowGraph cfg) {
+    checkCFG(cfg.getEntryBlock(), cfg);
+    super.postAnalyze(cfg);
+  }
+
+  private void checkCFG(SpecialBlock entry, ControlFlowGraph cfg) {
+
+    Set<Pair<Long, Set<AssignmentNode>>> visited = new HashSet<>();
+    visited.add(Pair.of(entry.getId(), Collections.emptySet()));
+
+    Deque<Pair<BlockImpl, Set<AssignmentNode>>> worklist = new ArrayDeque<>();
+    worklist.add(Pair.of((BlockImpl) entry, Collections.emptySet()));
+
+    while (!worklist.isEmpty()) {
+
+      Pair<BlockImpl, Set<AssignmentNode>> state = worklist.getLast();
+      worklist.removeLast();
+      List<Node> nodes = getBlockNodes(state.first);
+      Set<AssignmentNode> newDefs = new HashSet<>(state.second);
+
+      for (Node node : nodes) {
+        if (node instanceof AssignmentNode) {
+          Node lhs = ((AssignmentNode) node).getTarget();
+          Node rhs = ((AssignmentNode) node).getExpression();
+
+          if (lhs instanceof LocalVariableNode && hasAlwaysCall(lhs.getType())) {
+
+            // Reassignment to the lhs
+            if (isVarInPrevDefs(newDefs, (LocalVariableNode) lhs)) {
+              AssignmentNode latestAssignmentNode =
+                  getAssignmentNodeOfVar(newDefs, (LocalVariableNode) lhs);
+              reportAlwaysCallErrors(latestAssignmentNode, getStoreBefore(node), null);
+              newDefs.remove(latestAssignmentNode);
+            }
+
+            // If the rhs is ObjectCreationNode, MethodInvocationNode, MethodAccessNode or
+            // LocalVariableNode that exists in the newDefs (Note that if the localVariableNode
+            // exists in the newDefs it means it is assigned to a non-null values), then it adds
+            // the the AssignmentNode to the newDefs.
+            if ((rhs instanceof ObjectCreationNode)
+                || (rhs instanceof MethodInvocationNode)
+                || (rhs instanceof MethodAccessNode)
+                || (rhs instanceof LocalVariableNode
+                    && isVarInPrevDefs(newDefs, (LocalVariableNode) rhs))) {
+              newDefs.add((AssignmentNode) node);
+            }
+
+            // Ownership Transfer
+            if (rhs instanceof LocalVariableNode
```

Comment:
I would maybe handle this ownership transfer entirely within this `if`, rather than having the addition of the new assignment node in the previous `if`, for clarity